### DISCUSSION TITLE
Generating Dynamic Patterns

### DISCUSSION THREAD

**Discussion by Kingloko** [egraphs-good/egg#182] [Q&A] [answered]
Hey! 
I'll apologize in advance, I'm learning rust through using egg, so I might be missing something obvious. 

I can dynamically create an expression and add it to my egraph by following the examples from the documentation: 
```
//(Bar foo 1) 
    let foo_node = ToyLang::Symbol(Symbol::from("foo"));
    let foo = expr.add(foo_node);
    let one = expr.add(ToyLang::Num(1));
    let the_bar = expr.add(ToyLang::Bar([foo, one]));
```

What I'd like to do is create a pattern that matches `(Bar foo ?x)`, and I'm able to do that using `.parse().unwrap()`
But I'd like to programmatically generate this pattern, this is my "closest attempt" 
```
    let mut some_pat = PatternAst::default() // I picked PatternAst because it looks like its a RecExpr and there's a Pattern constructor that uses PatternAST 
    let some_var: = "?x".parse().unwrap();
    some_pat.add(ENodeOrVar::Var(some_var));
    some_pat.add(ENodeOrVar::ENode(&foo_node));
```

But I get the following error: ```the trait bound `&ToyLang: egg::Language` is not satisfied```.   

Any help would be a appreciated

**Comment by mwillsey** [comment, marked as answer]
I believe this is the error: 
```    
some_pat.add(ENodeOrVar::ENode(&foo_node));
```

should be 
```
some_pat.add(ENodeOrVar::ENode(foo_node));
```

Maybe with a `clone`. `ENodeOrVar::ENode` requires an owned enode.
